compiler/rustc_typeck/src/check/writeback.rs

```diff
@@ -223,7 +226,38 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
                 });
                 let index_ty = self.fcx.resolve_vars_if_possible(index_ty);
 
-                if base_ty.builtin_index().is_some() && index_ty == self.fcx.tcx.types.usize {
+                let mut lang_index = false;
+
+                let resolved_base_ty = self.resolve(*base_ty, &base.span);
+                let resolved_index_ty = self.resolve(index_ty, &index.span);
+                // (ouz-a #100498): Normally `[T] : std::ops::Index<usize>` should be normalized
+                // into [T] but currently `Where` clause stops the normalization process for it,
+                // here we check if the expr is `LangItem::Index` and index_ty is `usize` if so we don't
+                // `fix_index_builtin_expr`.
+                self.tcx().infer_ctxt().enter(|infcx| {
+                    let mut selection_context = SelectionContext::new(&infcx);
+
+                    let def_id = self.tcx().require_lang_item(LangItem::Index, Some(e.span));
+                    let substs = self.tcx().mk_substs([resolved_base_ty.into(), resolved_index_ty.into()].iter());
+                    let trait_ref = ty::TraitRef { def_id, substs };
+                    let binder = ty::Binder::dummy(trait_ref).without_const();
+                    let obligation = traits::Obligation::new(
+                        traits::ObligationCause::dummy(),
+                        self.fcx.param_env,
+                        binder,
+                    );
+                    let results = selection_context.select(&obligation);
+
+                    if let Ok(Some(ImplSource::Param(..))) = results && matches!(index_ty.kind(), Uint(ty::UintTy::Usize)){
+                        lang_index = true;
+                    }
+
+                });
+
+                if base_ty.builtin_index().is_some()
```

Comment:
When you pull that^ into its own method, we can actually do this check in the same method, and `return` if it's not true. That avoids having to do `selcx.select` if it's not actually `[T]`.